webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/policies/MergeNodeProperties.scala

```diff
@@ -0,0 +1,389 @@
+* the GNU General Public License version 3, the copyright holders add
+* the following Additional permissions:
+* Notwithstanding to the terms of section 5 (5. Conveying Modified Source
+* Versions) and 6 (6. Conveying Non-Source Forms.) of the GNU General
+* Public License version 3, when you create a Related Module, this
+* Related Module is not considered as a part of the work and may be
+* distributed under the license agreement of your choice.
+* A "Related Module" means a set of sources files including their
+* documentation that, without modification of the Source Code, enables
+* supplementary functions or services in addition to those offered by
+* the Software.
+*
+* Rudder is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder.services.policies
+
+import com.normation.rudder.domain.nodes.CompareProperties
+import com.normation.rudder.domain.nodes.GroupProperty
+import com.normation.rudder.domain.nodes.NodeGroupId
+import com.normation.rudder.domain.nodes.NodeInfo
+import com.normation.rudder.domain.policies.FullGroupTarget
+import com.normation.rudder.domain.policies.FullRuleTargetInfo
+import com.normation.rudder.domain.queries._
+import com.normation.errors._
+import com.normation.rudder.domain.policies.FullCompositeRuleTarget
+import com.normation.rudder.domain.policies.FullOtherTarget
+import GroupProp._
+import com.normation.rudder.domain.nodes.NodeProperty
+import com.softwaremill.quicklens._
+import net.liftweb.json.JsonAST.JNothing
+
+/**
+ * This file handle how node properties are merged with other (global, groups, etc)
+ * properties.
+ * Merge happens during policy generation and only spans that policy generation, ie nodes
+ * don't really have these properties: if you observe a node (for ex with API), you won't see them.
+ *
+ * Merge rules are the following:
+ * Overriding is authorized for same keys IF AND ONLY IF:
+ * - on a group by an other group, if the second group is a subgroup of the first,
+ *   i.e if he has a "AND" query composition, same nature (static/dynamic), and
+ *   a `SubGroupComparator` criterion with value the parent groupid.
+ * - on a node, if provider if the same.
+ *
+ * When overriding happens, we use a merge-override strategy, ie:
+ * - if the overriding value is a simple value or an array, it replaces the previous one
+ * - if the overriding value is an object and the overriden value is a simple type or an array,
+ *   the latter one is replaced by the former
+ * - if boeth overriding/overriden are an array, overriding values are appended to overriden array
```

Comment:
not sure here if we should only keep the overrding values

Reply:
now, we have both: `[overriden, overriding]`. I think it may not be what people want, but I also thing that whatever the choice, half people will want the other one